File: 5-Problem-A-Day/303. Range Sum Query - Immutable.java
class NumArray {

    int[] nums;
    
    public NumArray(int[] nums) {
        this.nums = nums;
    }
    
    public int sumRange(int i, int j) {
        int ans = 0;
        for (int k = i; k <= j; k++)
            ans += nums[k];
        return ans;
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(i,j);
 */

File: 2021/303. Range Sum Query - Immutable.cpp
class NumArray {
public:
    vector<int> preSum;
    NumArray(vector<int>& nums) {
        preSum.push_back(0);
        for (int n : nums) {
            preSum.push_back(preSum.back() + n);
        }
    }
    
    int sumRange(int left, int right) {
        return preSum[right + 1] - preSum[left];
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */

class NumArray {
public:
    vector<int> preSum;
    NumArray(vector<int>& nums) {
        preSum.push_back(0);
        for (int n : nums) {
            preSum.push_back(preSum.back() + n);
        }
    }
    
    int sumRange(int left, int right) {
        return preSum[right + 1] - preSum[left];
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */